### PR TITLE
Vault - funds lock meachanic

### PR DESCRIPTION
**Pull request by belbix**

### CODE REVIEW

File: contracts/base/vault/SmartVault.sol [hunk 1]
```diff
@@ -322,9 +398,39 @@ contract SmartVault is Initializable, ERC20Upgradeable, VaultStorage, Controllab
...
+        uint256 currentLockDuration = block.timestamp.sub(lockStart);
+        if (currentLockDuration < lockPeriod()) {
+          uint256 sharesBase = numberOfShares.mul(LOCK_PENALTY_DENOMINATOR - lockPenalty()).div(LOCK_PENALTY_DENOMINATOR);
+          uint256 toWithdraw = sharesBase.add(
+            numberOfShares.sub(sharesBase).mul(currentLockDuration).div(lockPeriod())
+          );
+          uint256 change = numberOfShares.sub(toWithdraw);
+          numberOfShares = toWithdraw;
+
+          _transfer(msg.sender, address(this), change);
```
Comment (olegnat) [inline contracts/base/vault/SmartVault.sol:426]: What will happen with those tokens? Could you please point me to the logic where those tokens will be used.

File: contracts/base/vault/SmartVault.sol [hunk 2]
```diff
@@ -322,9 +398,39 @@ contract SmartVault is Initializable, ERC20Upgradeable, VaultStorage, Controllab
...
+        userLockTs[msg.sender] = block.timestamp;
+      }
+      if (lockStart != 0 && lockStart < block.timestamp) {
+        uint256 currentLockDuration = block.timestamp.sub(lockStart);
+        if (currentLockDuration < lockPeriod()) {
+          uint256 sharesBase = numberOfShares.mul(LOCK_PENALTY_DENOMINATOR - lockPenalty()).div(LOCK_PENALTY_DENOMINATOR);
+          uint256 toWithdraw = sharesBase.add(
+            numberOfShares.sub(sharesBase).mul(currentLockDuration).div(lockPeriod())
+          );
+          uint256 change = numberOfShares.sub(toWithdraw);
```
Comment (olegnat) [inline contracts/base/vault/SmartVault.sol:423]: please name this variable more clearly.

File: contracts/infrastructure/PriceCalculator.sol [hunk 3]
```diff
@@ -233,7 +239,7 @@ contract PriceCalculator is Initializable, Controllable, IPriceCalculator {
     address pairAddress;
     // shortcut for firebird ice-weth
     // todo make more smart solution
-    if(_factory == 0x5De74546d3B86C8Df7FEEc30253865e1149818C8) {
+    if (_factory == 0x5De74546d3B86C8Df7FEEc30253865e1149818C8) {
```
Comment (olegnat) [inline contracts/infrastructure/PriceCalculator.sol:242]: Could you please extract this address to the constant with understandable name.

File: test/VaultUtils.ts [hunk 4]
```diff
@@ -128,8 +128,8 @@ export class VaultUtils {
   }
 
   public static async getVaultInfoFromServer() {
-    // return (await axios.get("https://api.tetu.io/api/v1/reader/vaultInfos?network=MATIC")).data;
-    return (await axios.get("http://localhost:8080/api/v1/reader/vaultInfos?network=MATIC")).data;
+    return (await axios.get("https://tetu-server-staging.herokuapp.com//api/v1/reader/vaultInfos?network=MATIC")).data;
+    // return (await axios.get("http://localhost:8080/api/v1/reader/vaultInfos?network=MATIC")).data;
```
Comment (olegnat) [inline test/VaultUtils.ts:132]: Please clean it up.

File: test/base/vault/DiamondVaultTest.ts [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+  beforeEach(async function () {
+    snapshotForEach = await TimeUtils.snapshot();
+  });
+
+  afterEach(async function () {
+    await TimeUtils.rollback(snapshotForEach);
+  });
+
+
+  it("check locked funds in loop", async () => {
```
Comment (olegnat) [inline test/base/vault/DiamondVaultTest.ts:73]: This test is too generic, I can't pick up testing scenarios. Could you please divide it for small isolated use cases.

File: contracts/base/vault/SmartVault.sol (hunk 1 above)
Comment (belbix) [reply contracts/base/vault/SmartVault.sol:422]: it will be used as rewards

File: test/base/vault/DiamondVaultTest.ts (hunk 5 above)
Comment (belbix) [reply test/base/vault/DiamondVaultTest.ts:73]: some functionality can't be tested without multiple interactions with the vault
it requires a loop with users actions that probably can affect some logic

File: contracts/base/interface/ISmartVault.sol [hunk 6]
```diff
@@ -86,13 +90,23 @@ interface ISmartVault {
...
 
   function stop() external;
 
   function ppfsDecreaseAllowed() external view returns (bool);
+
+  function lockPeriod() external view returns (uint256);
+
+  function lockPenalty() external view returns (uint256);
+
+  function lockAllowed() external view returns (bool);
```
[the remark this answers is not in the document]
Comment (belbix) [reply contracts/base/interface/ISmartVault.sol:111]: this variably only allowed lock mechanic but not switch it on
for enabling locks you should setup lock period and lock initial penalty

**Pull request merged**